[Charlton Teo] iP

bigcrushes opened the pull request:
### DukePro 

> “Your mind is for having ideas, not holding them.” – David Allen

DukePro frees your mind of having to remember things you need to do. It's,
- Text-based
- Easy to learn
- ~~fast~~ SUPER FAST to use

All you need to do is,
1. download it from [here](https://github.com/bigcrushes/ip)
2. double click it
3. add your tasks
4. let it manage your tasks for you :yum:

And it is **FREE**
Features:
- [ ] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

bigcrushes committed:
Level-1

bigcrushes committed:
Level-2

bigcrushes committed:
added mark and unmark

bigcrushes committed:
added todo, deadline, event classes

bigcrushes committed:
added automated text ui testing

bigcrushes committed:
added exceptions for empty descriptions and unknown commands

bigcrushes committed:
added delete function

bigcrushes committed:
Allow storing of tasks on harddrive

bigcrushes committed:
Add date time conversion functionality for deadlines and event timings

bigcrushes committed:
Tidy up imports

bigcrushes committed:
Tidy up imports

bigcrushes committed:
Merge branch 'master' of https://github.com/bigcrushes/ip

bigcrushes committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/Duke.java

bigcrushes committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

bigcrushes committed:
Fix merge conflicts

bigcrushes committed:
Refactor code for more OOP

bigcrushes committed:
Refactor code for more OOP

bigcrushes committed:
Merge branch 'add-gradle-support'

bigcrushes committed:
Add gradle support

bigcrushes committed:
Add JUnit tests

bigcrushes committed:
Fix minor bugs

bigcrushes committed:
Add javadocs and tweek code to comply with coding standards

bigcrushes committed:
Add find command

bigcrushes committed:
Add javadocs

bigcrushes committed:
Tweek code to comply with coding standard

bigcrushes committed:
Add find command to find tasks with keyword

bigcrushes committed:
Merge branch 'branch-A-CodingStandard'

bigcrushes committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/Ui.java

bigcrushes committed:
Deconflict merge conflicts

bigcrushes committed:
Merge branch 'master' of https://github.com/bigcrushes/ip

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/TaskList.java

bigcrushes committed:
Add GUI for chatbot

bigcrushes committed:
Align code with java coding standards using checkstyle

carriezhengjr commented on src/main/java/duke/DialogBox.java:
```diff
@@ -0,0 +1,61 @@
+import javafx.scene.image.ImageView;
+import javafx.scene.layout.HBox;
+
+/**
+ * An example of a custom control using FXML.
+ * This control represents a dialog box consisting of an ImageView to represent the speaker's face and a label
+ * containing text from the speaker.
+ */
```
Very detailed description of the class! 👍

carriezhengjr commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,30 @@
+
+    /**
+     * Initialises Duke chatbot
+     * @throws IOException
```
Perhaps you could add a short and brief description for the exception here. 🤔

carriezhengjr commented on src/main/java/duke/MainWindow.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    @FXML
+    private void handleUserInput() throws IOException {
```
Perhaps you could add a JavaDoc comment for this method. 🤔

carriezhengjr commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,117 @@
+    public String executeCommand(TaskList tasks, Storage storage) throws IOException {
+        String[] commands = command.strip().split(" ", 2);
+        try {
+            switch (commands[0]) {
```
Good distinction for String[] commands as a multi-valued variable. 👍 
Very very small nitpick, you could name commands[0] as e.g. commandType to improve readability to show what commands[0] represents, but just personal preference. 😄

carriezhengjr commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,74 @@
+        this.filename = filename;
+    }
+
+    /**
+     * Loads list of tasks previously saved in hard disk
+     * If no previous file exists, create a new file
+     * @return ArrayList of tasks that were loaded from hard disk
+     * @throws IOException
+     */
```
There could be an empty line between description and parameter section as suggested here: https://se-education.org/guides/conventions/java/intermediate.html#comments
Same for other JavaDoc comments. 😃

carriezhengjr commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,74 @@
+            default:
+                break;
+            }
+            if (loadedTask[1].equals("X")) {
```
Perhaps you could name loadTask[0] and loadTask[1], e.g. taskType and isDone or hasMarked, to improve readability. 🤔 😃

carriezhengjr commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,100 @@
+
+    /**
+     * Adds a task to the task list
+     * @param task current task list
```
Is it current task list, or current task? 🤔

carriezhengjr commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,83 @@
+/**
+ * Class to handle all inputs and outputs for user interaction
+ */
+public class Ui {
```
Overall, clear documentation and comprehensive code that is easy to read! 👍

carriezhengjr commented on src/main/java/duke/tasks/Deadline.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    /**
+     * {@inheritDoc}
```
Good use of @inheritDoc to reuse the header comment from the parent method! 👍

carriezhengjr reviewed (commented):
Overall, well-documented and comprehensive code 👍  
There are just some very small nits on code quality and coding standard. 😃

nicolelim02 commented on src/main/java/duke/Main.java:
```diff
@@ -0,0 +1,34 @@
+
+    private Duke duke = new Duke();
+
+    public Main() throws IOException {
+    }
```
Is this an unused method? 🤔

ShenyiCui replied:
Please remove empty methods

nicolelim02 commented on src/main/java/duke/MainWindow.java:
```diff
@@ -0,0 +1,60 @@
+    private void handleUserInput() throws IOException {
+        String input = userInput.getText();
+        String response = duke.getResponse(input);
+        dialogContainer.getChildren().addAll(
+                DialogBox.getUserDialog(input, userImage),
+                DialogBox.getDukeDialog(response, dukeImage)
+        );
```
Good use of line breaks!

nicolelim02 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,117 @@
+            case "list":
+                return ui.getCorrectMessage(Ui.Commands.LIST, tasks, 0);
+            case "mark":
+                if (commands.length == 1) {
+                    throw new EmptyDescriptionException("mark");
```
Perhaps you can write a method for this so you don't have to repeat it for other commands?

nicolelim02 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,117 @@
+                if (eventCommands.length == 1) {
+                    return ui.getErrorMessage("datetime");
+                }
+                try {
+                    LocalDate timeStringParsed = LocalDate.parse(eventCommands[1]);
+                    tasks.add(new Event(eventCommands[0],
+                            timeStringParsed.format(DateTimeFormatter.ofPattern("MMM dd yyyy"))));
+                    return ui.getCorrectMessage(Ui.Commands.TASK, tasks, tasks.getSize() - 1);
+                } catch (DateTimeParseException error) {
+                    return ui.getErrorMessage("datetime");
```
Perhaps you can also write a separate method to check for the validity of a date provided by the user, as stated in this [section](https://nus-cs2103-ay2223s1.github.io/website/se-book-adapted/chapters/codeQuality.html#minimize-code-duplication) of the textbook

nicolelim02 reviewed (commented):
Overall, your code is easy to understand. Keep up the good work! 😀

ShenyiCui commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,30 @@
+package duke;
+
+
```
It would be better if you kept the spaces between your package and imports to just 1 white line instead of 2.

ShenyiCui commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,117 @@
+     * @return boolean true if bye command entered and chatbot should be stopped, false otherwise
+     */
+    public String executeCommand(TaskList tasks, Storage storage) throws IOException {
+        String[] commands = command.strip().split(" ", 2);
```
you can consider writing
```Java
command.strip()
       .split(" ", 2);
```

ShenyiCui commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,117 @@
+                tasks.add(new Todo(commands[1]));
+                return ui.getCorrectMessage(Ui.Commands.TASK, tasks, tasks.getSize() - 1);
+            case "event":
+                if (commands.length == 1) {
```
This entire switch case statement is a little bit unreadable, I would consider adding more vertical spacing and seeing where you can abstract out some of your logic. 

for example, does it make sense for you to do your error check and throwing here? or should they be abstracted into your classes? `Ask Don't Tell`.

ShenyiCui reviewed (changes requested):
Lgtm, nearly, just a few things I would change for improved code readability.

bigcrushes committed:
Improve code quality

bigcrushes committed:
Add assertions

bigcrushes committed:
Add assertion

bigcrushes committed:
Merge pull request #1 from bigcrushes/branch-A-Assertions

Add assertions

bigcrushes committed:
Merge branch 'master' into branch-A-CodeQuality

bigcrushes committed:
Merge pull request #2 from bigcrushes/branch-A-CodeQuality

Improve code quality

bigcrushes committed:
Improve find

Find command only searches in task description

A find command that is able to search for dates can add more
functionality for the user.

Let's,
* update find algorithm to find text in dates
* able to find words in any of the tasks' tags in future

bigcrushes committed:
Add tagging system

No tagging system in place currently.

A tagging system will allow users to search
their tasks by customisable tags.

Let's,
* implement a tagging system
* let find command search tags too

bigcrushes committed:
Update README.md

bigcrushes committed:
Add files via upload

bigcrushes committed:
Fix minor coding standard problems

bigcrushes committed:
Merge branch 'master' of https://github.com/bigcrushes/ip

bigcrushes committed:
Add user guide

bigcrushes committed:
Improve user guide

bigcrushes committed:
Add better formatting to user guide

bigcrushes committed:
Add command summary to user guide
